File: 0773-sliding-puzzle/0773-sliding-puzzle.py
from collections import deque

class Solution:
    def __init__(self):
        self.directions = [
            [1, 3],     # 0
            [0, 2, 4],  # 1
            [1, 5],     # 2
            [0, 4],     # 3
            [1, 3, 5],  # 4
            [2, 4]      # 5
        ]

    def slidingPuzzle(self, board):
        target = "123450"
        start = "".join(str(cell) for row in board for cell in row)
        visited = {start: 0}
        queue = deque([(start.index('0'), start)])

        while queue:
            zero_pos, current_state = queue.popleft()
            if current_state == target:
                return visited[current_state]

            for neighbor in self.directions[zero_pos]:
                new_state = list(current_state)
                new_state[zero_pos], new_state[neighbor] = new_state[neighbor], new_state[zero_pos]
                new_state_str = "".join(new_state)
                if new_state_str not in visited:
                    visited[new_state_str] = visited[current_state] + 1
                    queue.append((neighbor, new_state_str))

        return -1

File: 0773-sliding-puzzle/0773-sliding-puzzle.cpp
class Solution {
public:
    vector<vector<int>>dir = {
        {1,3}, // 1
        {0,2,4}, // 2
        {1,5}, // 3
        {0,4}, // 4
        {1,3,5},// 0
        {2,4}// 5
    };
    unordered_map<string,int>vis;

    int BFS(int p0,string s){
        queue< pair<int,string>>q;
        q.push({p0,s});
        vis[s] = 0;
        while(q.size()){
            int zeroPos = q.front().first;
            string s = q.front().second; // 412503 | 0 

            q.pop();
            if(s == "123450")return vis["123450"];
            for(int nInd :dir[zeroPos]){
                string curState = s;
                swap(curState[zeroPos],curState[nInd]);
                if(!vis.count(curState) ){
                    vis[curState] = vis[s]+1;
                    q.push({nInd,curState});
                }
            }

        }
        return -1;
    }
    int slidingPuzzle(vector<vector<int>>& board) {
        string s ;
        for(auto row : board)
            for(auto c : row)
                s+='0'+c;

        return BFS(s.find('0'),s);

    }
};